[bug/enhancement] Multiple fixes / extra formatting

TJulesL opened the pull request:
Added :

- Font loading, code now can load fonts from a string. In old code it wasn't clear how to load fonts or if you could load them at all. Also added an check that when an error occurs it prints it and uses the system default font

- Fixed borders for circles / rectangles, It didnt work at all at first because there was no logic implemented. I have implemented it so it draws for example a big circle (border) with a little circle within (inner circle).

- play.new_text now gives an specific error when the input is not a string

- fixed images, you are now able to load images in with a location on the computer as string 

- Added some extra "documentation" in generator.py

- Added a new feature so you can smooth out the corners with a given "border_radius". I will implement functionality later so it also works for rectangles without any borders, since this is not yet possible at the moment

koen1711 commented:
Hi thank you so much for your suggestions and fixes, we are currently heavily working on the documentation pages and adding more tests etc. Let me know if you find anymore bugs so those can be addressed

koen1711 commented:
I'll review all of this tonight and then let you know what I think, in my big refactor one or two of the logic might have became unimplemented which in the future won't happen again of course. Since that would be devastating for the classroom

TJulesL commented:
Is goed neem je tijd, als project ga ik nu ook onder begeleiding van meneer postma werken (als het goed is) dus ik zal waarschijnlijk nog wel vaker een paar pull requests of eventuele features zal committen (achtergrondmuziek / sound effects implementeren etc).

koen1711 commented on play/api/generators.py:
```diff
@@ -10,16 +10,16 @@
+    words: str = "",
+    x: int = 0,
+    y: int = 0,
+    font: str = "C:/location/to/font_file.ttf",
```
We use None for a reason here to load, the default font. Since not everyone uses windows I would recommend using a more general file path like /path/to/file

koen1711 commented on play/api/generators.py:
```diff
@@ -10,16 +10,16 @@
+    angle: int = 0,
+    transparency: int = 100,
+    size: int = 100,
+) -> object:
```
-> _Text

koen1711 commented on play/api/generators.py:
```diff
@@ -46,17 +49,18 @@ def new_text(  # pylint: disable=too-many-arguments
+    angle: int = 0,
+    transparency: int = 100,
+    size: int = 100,
+) -> object:
```
_Box

koen1711 commented on play/api/generators.py:
```diff
@@ -78,23 +83,24 @@ def new_box(  # pylint: disable=too-many-arguments
+    transparency: int = 100,
+    size: int = 100,
+    angle: int = 0,
+) -> object:
```
_Circle

koen1711 commented on play/api/generators.py:
```diff
@@ -121,17 +127,17 @@ def new_circle(  # pylint: disable=too-many-arguments
+    y1 = None,
+    transparency: int = 100,
+    size: int = 100,
+) -> object:
```
_Line

koen1711 commented on play/api/generators.py:
```diff
@@ -160,8 +166,13 @@ def new_line(  # pylint: disable=too-many-arguments
+    size: int = 100,
+    angle: int = 0, 
+    transparency: int = 100
+) -> object:  # pylint: disable=too-many-arguments
```
_Image

koen1711 commented:
Please also fix your code using `black ./play ./tests` and `pylint ./play`

koen1711 commented:
Also, before merging. Please change the version in setup.py using semantic versioning

TJulesL commented:
I have changed the version, a minor version increment because of the added features/fixes. Also changed the text.py file to now use logging.py. At last i also used black / pylint to format it

TJulesL commented:
i just noticed a fatal error in the code which glitches out the objects (ball and box) so **Please dont merge just yet**. I will try and fix it right now

TJulesL commented:
i just noticed a fatal error in the code which glitches out the objects (ball and box) so **Please dont merge just yet**. I will try and fix it right now

TJulesL commented:
> > i just noticed a fatal error in the code which glitches out the objects (ball and box) so **Please dont merge just yet**. I will try and fix it right now
> 
> Alright, could you show the code where that happens? Because if you update box.y it's logical it doesn't work since you do it yourself.



So the problem wasn't my code which i thought it would be (something like the borders i implemented), it was that there was no code for drawing the rotation of boxes and circles. This made it so that pymunk calculated the collision correctly however the box was never drawn on that angle. The reason why it looked like a piece of the box was "under the screen" was because the x/y coords for the middle were calculated with the context of the rectangle laying on its side. This also explains the weird collision in parts where the box wasn't. 

I coded the logic for it and fixed it + added it in one of my recent commits where i formatted it with black and pylint. Also tested it again, collision with different types of objects work fine now + the rounding of borders etc.
